nni/compression/pytorch/utils/counter.py

```diff
@@ -26,110 +225,113 @@ def count_flops_params(model: nn.Module, input_size, custom_ops=None, verbose=Tr
     ---------
     model : nn.Module
         target model.
-    input_size: list, tuple
-        the input shape of data
+    x: tuple or tensor
+        the input shape of data or a tensor as input data
     custom_ops: dict
         a mapping of (module: custom operation)
         the custom operation will overwrite the default operation.
         for reference, please see ``custom_mask_ops``.
+    verbose: bool
+        If False, mute detail information about modules. Default is True.
+    mode:
+        the mode of how to collect information. If the mode is set to `default`,
+        only the information of convolution and linear will be collected.
+        If the mode is set to `full`, other operations will also be collected.
 
     Returns
     -------
     flops: float
         total flops of the model
     params:
         total params of the model
+    results: dict
+        detail information of modules
     """
 
-    assert input_size is not None
+    assert isinstance(x, tuple) or isinstance(x, torch.Tensor)
+    assert mode in ['default', 'full']
+
+    original_device = next(model.parameters()).device
+    training = model.training
 
-    device = next(model.parameters()).device
-    inputs = torch.randn(input_size).to(device)
+    if torch.is_tensor(x[0]):
+        x = (t.to(original_device) for t in x)
+    else:
+        x = (torch.zeros(x).to(original_device), )
 
-    hook_module_list = []
+    handler_collection = []
+    results = dict()
     if custom_ops is None:
         custom_ops = {}
-    custom_mask_ops.update(custom_ops)
-    prev_m = None
-    for m in model.modules():
-        weight_mask = None
-        m_type = type(m)
-        if m_type in custom_mask_ops:
-            if isinstance(prev_m, PrunerModuleWrapper):
-                weight_mask = prev_m.weight_mask
-
-            m.register_buffer('weight_mask', weight_mask)
-            hook_module_list.append(m)
-        prev_m = m
-
-    flops, params = profile(model, inputs=(inputs, ), custom_ops=custom_mask_ops, verbose=verbose)
+    register_ops = mode_cls[mode]
+    register_ops.update(custom_ops)
 
+    # set leaf module name
+    prev_m = None
+    for name, m in model.named_modules():
+        if len(list(m.children())) > 0:  # skip for non-leaf module
+            continue
+        setattr(m, "_name", name)
```

Comment:
what is this line used for?

Reply:
to set a name for the module, so the name can be used in the forward hook